rust/sdk/src/client.rs

```diff
@@ -388,6 +316,356 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+    Empty,
+}
+
+impl<'a> Publish<'a> {
+    fn new(client: &'a ActyxClient) -> Self {
+        Self::Initial {
+            client,
+            request: PublishRequest { data: vec![] },
+        }
+    }
+
+    // TODO: add an example showing the subsequent calls instead of a sentence
+    /// Add an event for publishing.
+    ///
+    /// Subsequent calls will add the events rather than replacing them.
+    pub fn event<E: Serialize>(mut self, tags: TagSet, event: &E) -> Result<Self, serde_cbor::Error> {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.data.push(PublishEvent {
+                tags,
+                payload: Payload::compact(event)?,
+            });
+        }
+        Ok(self)
+    }
+
+    // TODO: add an example showing the subsequent calls instead of a sentence
+    /// Add events for publishing, from an iterable.
+    ///
+    /// Subsequent calls will add the events rather than replacing them.
+    pub fn events<E: IntoIterator<Item = impl Into<PublishEvent>>>(mut self, events: E) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.data.extend(events.into_iter().map(Into::into));
+        }
+        self
+    }
+}
+
+impl<'a> Future for Publish<'a> {
+    type Output = anyhow::Result<PublishResponse>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> std::task::Poll<Self::Output> {
+        let this = self.get_mut();
+        loop {
+            *this = match replace(this, Publish::Empty) {
+                Publish::Initial { client, request } => {
+                    let publish_response = async move {
+                        let publish_url = client.events_url("publish");
+                        let response = client.do_request(|c| c.post(publish_url).json(&request)).await?;
+                        let body = response.bytes().await?;
+                        Ok(serde_json::from_slice::<PublishResponse>(&body)?)
+                    };
+                    Publish::Pending(publish_response.boxed())
+                }
+                Publish::Pending(mut publish_response_future) => {
+                    let polled = publish_response_future.poll_unpin(cx);
+                    if polled.is_pending() {
+                        *this = Publish::Pending(publish_response_future);
+                    }
+                    return polled;
+                }
+                Publish::Empty => panic!("Polling a terminated Publish future"),
+            };
+        }
+    }
+}
+
+impl<'a> FusedFuture for Publish<'a> {
+    fn is_terminated(&self) -> bool {
+        if let Publish::Empty = self {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+pub enum Query<'a> {
+    Initial {
+        client: &'a ActyxClient,
+        request: QueryRequest,
+    },
+    Pending(BoxFuture<'a, anyhow::Result<BoxStream<'a, QueryResponse>>>),
+    Empty,
```

Comment:
A nitpick:
`Spent`, `Depleted`, or `Exhausted` can give more meaning instead of `Empty`.
My initial reaction to Empty is that it felt like the initial state rather than the end.

Reply:
Since this is also used as an intermediate value while polling, `Void` would be more appropriate.